context-propagation/src/main/java/io/smallrye/mutiny/context/BaseContextPropagationInterceptor.java

```diff
@@ -0,0 +1,392 @@
+package io.smallrye.mutiny.context;
+
+import java.util.concurrent.Executor;
+import java.util.function.*;
+
+import io.smallrye.context.SmallRyeThreadContext;
+import io.smallrye.context.impl.Contextualized;
+import io.smallrye.mutiny.infrastructure.CallbackDecorator;
+import io.smallrye.mutiny.tuples.Functions;
+
+public abstract class BaseContextPropagationInterceptor implements CallbackDecorator {
+
+    /**
+     * Gets the Context Propagation ThreadContext. External
+     * implementations may implement this method.
+     *
+     * @return the ThreadContext
+     */
+    protected abstract SmallRyeThreadContext getThreadContext();
+
+    @Override
+    public <T> Supplier<T> decorate(Supplier<T> supplier) {
+        SmallRyeThreadContext context = getThreadContext();
+        if (context.isContextualized(supplier) || context.isEmpty()) {
```

Comment:
To be clear, this is already done in `contextualSupplier` so you don't need to do it. It's only useful if you wrap the wrappers.

Reply:
do you mean the `isEmpty` or the `isContextualized` or both ?

Reply:
Both, we already check for both in `contextualConsumer`. This is only useful if you create a lambda when you call `contextualConsumer` because it will save you creating that lambda.

Reply:
Oh hold on you're right, you cannot get rid of `isContextualized` because the spec forces us to throw if it's true and you don't want that. We can make our `contextualConsumerUnlessContextualized` method public if you want, or you keep on doing this.

Reply:
I'm good with the current version. I have enough tests to verify _all_ the cases (with and without context, already contextualized or not).